File: apps/indexer-sol/src/units.ts
import {
  NATIVE_DECIMALS,
  PUMP_FEEL_DEFAULTS,
  SPOT_PRICE_TOKEN_UNIT,
} from "@pump/solana-sdk";

/** Format integer base units as a decimal string (no scientific notation). */
export function formatUnits(value: bigint, decimals: number): string {
  if (decimals < 0) throw new Error("decimals must be >= 0");
  const neg = value < 0n;
  const v = neg ? -value : value;
  const base = 10n ** BigInt(decimals);
  const whole = v / base;
  const frac = (v % base).toString().padStart(decimals, "0").replace(/0+$/, "");
  const body = frac.length > 0 ? `${whole}.${frac}` : `${whole}`;
  return neg ? `-${body}` : body;
}

export function lamportsToSol(lamports: bigint): string {
  return formatUnits(lamports, NATIVE_DECIMALS);
}

export function tokenAmountToDecimal(
  amount: bigint,
  decimals: number = PUMP_FEEL_DEFAULTS.tokenDecimals
): string {
  return formatUnits(amount, decimals);
}

/** SOL per whole token from trade amounts. */
export function executionPriceSol(
  solLamports: bigint,
  tokenAmount: bigint,
  tokenDecimals: number = PUMP_FEEL_DEFAULTS.tokenDecimals
): string {
  if (tokenAmount === 0n) return "0";
  // (sol/1e9) / (tok/10^d) = sol * 10^d / (tok * 1e9)
  const scale = 18n;
  const numer = solLamports * 10n ** BigInt(tokenDecimals) * 10n ** scale;
  const denom = tokenAmount * 10n ** BigInt(NATIVE_DECIMALS);
  const scaled = numer / denom;
  return formatUnits(scaled, Number(scale));
}

/**
 * On-chain spot_price = (poolSolLamports * TOKEN_UNIT_9) / poolTokenBase.
 * Convert to SOL per whole token (10^tokenDecimals base units).
 */
export function spotPriceSolPerToken(
  spotPriceU64: bigint,
  tokenDecimals: number = PUMP_FEEL_DEFAULTS.tokenDecimals
): string {
  if (spotPriceU64 === 0n) return "0";
  // SOL/whole = spot * 10^d / (TOKEN_UNIT_9 * 1e9)
  const scale = 18n;
  const numer =
    spotPriceU64 * 10n ** BigInt(tokenDecimals) * 10n ** scale;
  const denom = SPOT_PRICE_TOKEN_UNIT * 10n ** BigInt(NATIVE_DECIMALS);
  const scaled = numer / denom;
  return formatUnits(scaled, Number(scale));
}

export function startingSpotFromVirtual(
  virtualSolLamports: bigint,
  virtualTokenReserve: bigint,
  tokenDecimals: number = PUMP_FEEL_DEFAULTS.tokenDecimals
): string {
  if (virtualTokenReserve === 0n) return "0";
  return executionPriceSol(virtualSolLamports, virtualTokenReserve, tokenDecimals);
}

/** FDV-style mcap in SOL for 1B whole tokens (matches EVM indexer * 1e9). */
export function marketCapSolFromSpot(spotSolPerToken: string): string {
  const price = Number(spotSolPerToken);
  if (!Number.isFinite(price) || price <= 0) return "0";
  return String(price * 1_000_000_000);
}

export function asBigInt(v: unknown): bigint {
  if (typeof v === "bigint") return v;
  if (typeof v === "number") return BigInt(Math.trunc(v));
  if (typeof v === "string") return BigInt(v);
  throw new Error(`expected bigint-compatible, got ${typeof v}`);
}

export function asString(v: unknown): string {
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "bigint") return String(v);
  throw new Error(`expected string-compatible, got ${typeof v}`);
}

export function asBool(v: unknown): boolean {
  return Boolean(v);
}

export function eventId(signature: string, logIndex: number): string {
  return `${signature}:${logIndex}`;
}

export function feeSplitKey(signature: string, mint: string): string {
  return `${signature}:${mint}`;
}

File: apps/indexer-sol-go/internal/units/units.go
package units

import (
	"fmt"
	"math/big"
	"strings"
)

const LamportsPerSol = 1_000_000_000

// On-chain spot uses TOKEN_UNIT_9 base (see @pump/solana-sdk SPOT_PRICE_TOKEN_UNIT).
const SpotPriceTokenUnit = 1_000_000_000

func LamportsToSol(lamports uint64) string {
	if lamports == 0 {
		return "0"
	}
	num := new(big.Rat).SetUint64(lamports)
	den := new(big.Rat).SetUint64(LamportsPerSol)
	out := new(big.Rat).Quo(num, den)
	return out.FloatString(9)
}

func TokenAmountToDecimal(amount uint64, decimals int) string {
	if amount == 0 {
		return "0"
	}
	if decimals <= 0 {
		return fmt.Sprintf("%d", amount)
	}
	num := new(big.Rat).SetUint64(amount)
	den := new(big.Rat).SetInt(exp10(decimals))
	out := new(big.Rat).Quo(num, den)
	return out.FloatString(decimals)
}

// SpotPriceSolPerToken converts on-chain spot (lamports per TOKEN_UNIT_9 base units)
// to SOL per whole token — must match apps/indexer-sol/src/units.ts spotPriceSolPerToken.
func SpotPriceSolPerToken(spotRaw uint64, decimals int) string {
	if spotRaw == 0 {
		return "0"
	}
	if decimals < 0 {
		decimals = 6
	}
	const scale = 18
	num := new(big.Int).SetUint64(spotRaw)
	num.Mul(num, exp10(decimals))
	num.Mul(num, exp10(scale))
	den := new(big.Int).SetUint64(SpotPriceTokenUnit)
	den.Mul(den, exp10(9))
	scaled := new(big.Int).Quo(num, den)
	return formatScaledInt(scaled, scale)
}

func ExecutionPriceSol(solAmount, tokenAmount uint64, decimals int) string {
	tokens := TokenAmountToDecimal(tokenAmount, decimals)
	if tokenAmount == 0 {
		return "0"
	}
	sol := LamportsToSol(solAmount)
	solRat, _ := new(big.Rat).SetString(sol)
	tokRat, _ := new(big.Rat).SetString(tokens)
	if tokRat.Sign() == 0 {
		return "0"
	}
	out := new(big.Rat).Quo(solRat, tokRat)
	return out.FloatString(12)
}

func MarketCapSolFromSpot(spotPrice string) string {
	const totalSupply = 1_000_000_000
	spot, ok := new(big.Rat).SetString(spotPrice)
	if !ok || spot.Sign() <= 0 {
		return "0"
	}
	out := new(big.Rat).Mul(spot, new(big.Rat).SetInt64(totalSupply))
	return out.FloatString(6)
}

func EventID(signature string, logIndex int) string {
	return fmt.Sprintf("%s:%d", signature, logIndex)
}

func FeeSplitKey(signature, mint string) string {
	return signature + ":" + mint
}

func exp10(n int) *big.Int {
	return new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(n)), nil)
}

func formatScaledInt(v *big.Int, decimals int) string {
	if v.Sign() == 0 {
		return "0"
	}
	neg := v.Sign() < 0
	abs := new(big.Int).Abs(v)
	base := exp10(decimals)
	whole := new(big.Int).Quo(abs, base)
	frac := new(big.Int).Mod(abs, base)
	fracStr := fmt.Sprintf("%0*s", decimals, frac.String())
	fracStr = strings.TrimRight(fracStr, "0")
	if fracStr == "" {
		if neg {
			return "-" + whole.String()
		}
		return whole.String()
	}
	body := whole.String() + "." + fracStr
	if neg {
		return "-" + body
	}
	return body
}
